python/ray/serve/tests/test_target_capacity.py

```diff
@@ -373,6 +380,358 @@ def test_autoscaling_scale_to_zero(
     )
 
 
+@serve.deployment(ray_actor_options={"num_cpus": 0})
+class ControlledLifecycleDeployment:
+    async def __init__(self):
+        self.lifecycle_signal = ray.get_actor(
+            name="lifecycle_signal", namespace=SERVE_NAMESPACE
+        )
+        await self.lifecycle_signal.wait.remote()
+
+        self.request_signal = ray.get_actor(
+            name="request_signal", namespace=SERVE_NAMESPACE
+        )
+
+    async def __call__(self) -> str:
+        try:
+            await self.request_signal.wait.remote()
+        except RayActorError:
+            print("Request signal actor already dead. Running request.")
+        return "Hello world!"
+
+    async def __del__(self):
+        # Use try-except here to avoid cluttering logs if the SignalActor
+        # gets torn down before the replicas during clean up.
+        try:
+            await self.lifecycle_signal.wait.remote()
+        except RayActorError:
+            print("Lifecycle signal actor already dead. Replica exiting.")
+
+
+def create_controlled_app(config: Dict) -> Application:
+    num_replicas = config["num_replicas"]
+    return ControlledLifecycleDeployment.options(
+        name="controlled",
+        num_replicas=num_replicas,
+    ).bind()
+
+
+def create_autoscaling_controlled_app(config: Dict) -> Application:
+    min_replicas = config["min_replicas"]
+    initial_replicas = config["initial_replicas"]
+    max_replicas = config["max_replicas"]
+    return ControlledLifecycleDeployment.options(
+        name="controlled",
+        autoscaling_config=dict(
+            min_replicas=min_replicas,
+            initial_replicas=initial_replicas,
+            max_replicas=max_replicas,
+            target_num_ongoing_requests_per_replica=1,
+            metrics_interval_s=0.01,
+            look_back_period_s=0.01,
+            upscale_delay_s=0.01,
+            downscale_delay_s=0.01,
+        ),
+        graceful_shutdown_timeout_s=0,
+    ).bind()
+
+
+class TestTargetCapacityUpdateAndServeStatus:
+    def check_num_running_replicas(
+        self, expected_num_running_replicas: int, app_name: str, deployment_name: str
+    ) -> bool:
+        deployment = serve.status().applications[app_name].deployments[deployment_name]
+        num_running_replicas = deployment.replica_states.get(ReplicaState.RUNNING, 0)
+        assert num_running_replicas == expected_num_running_replicas
+        return True
+
+    def apply_config_and_check_status(
+        self,
+        client: ServeControllerClient,
+        target_capacity: Optional[float],
+        config: ServeDeploySchema,
+        app_name: str,
+        deployment_name: str,
+        check_statuses: bool = True,
+    ):
+        """Applies config with specified target_capacity."""
+
+        config.target_capacity = target_capacity
```

Comment:
make a copy to avoid mutating passed-in arg?

Reply:
The test now `deepcopies` the `config` before mutating it.